### PR TITLE
CLOUDP-165325: Data Federation and private endpoint Atlas CLI Changes

### PR DESCRIPTION
**Pull request by cveticm**
## Proposed changes

Adding generation of Data Federation for `kubernetes config generate` command.
New flag `--dataFederationName` has been added to export specified data federation instances, similar to flag `--clusterName`
Data Federation Private Endpoints will not be included in this due to the inability to identify what resource a private endpoint is connected to.

_Jira ticket:_ CLOUDP-#165325
[https://jira.mongodb.org/browse/CLOUDP-165325](url)

Closes #[issue number]

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation in document requirements section listed in [CONTRIBUTING.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/CONTRIBUTING.md) (if appropriate)
- [ ] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
- [ ] I have updated [test/README.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/test/README.md) (if an e2e test has been added)
- [ ] I have run `make fmt` and formatted my code

### CODE REVIEW

**Review by roothorp — APPROVED** [review]
Awesome - love all the tests!

**Review by josvazg — DISMISSED** [review]
LGTM
Good job!

File: internal/kubernetes/operator/config_exporter.go
Comment (gssbzn) [inline internal/kubernetes/operator/config_exporter.go:283]:
```suggestion
func (e *ConfigExporter) fetchDataFederationNames() ([]string, error) {
```

File: internal/kubernetes/operator/datafederation/datafederation_test.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation_test.go:41]:
```suggestion
const (
projectName = "testProject-1"
dataFederationName = "testDataFederation-1"
 targetNamespace = "test-namespace-1"
 resourceVersion = "x.y.z"
 projectID = "test-project-id"
)
```

File: internal/kubernetes/operator/datafederation/datafederation_test.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation_test.go:205]:
```suggestion
			t.Logf("E:%s\r\n; G:%s\r\n", expJs, gotJs)
```

File: internal/store/data_federation.go
Comment (gssbzn) [inline internal/store/data_federation.go:18]: any reason not to use the new SDK? we should stop using this package

File: test/e2e/atlas/helper_test.go
Comment (gssbzn) [inline test/e2e/atlas/helper_test.go:786]: this could break in the future, these helpers rely on the CLI since the CLI will abstract the env var part and in the future once we remove mongocli from this repo we will stop supporting these envs, if you guys will add this you guys are in the hook for the support latter on cc @igor-karpukhin

Comment (gssbzn) [reply test/e2e/atlas/helper_test.go:786]: also, and this will be a hard line, use the new SDK if we are keeping this as it is

Comment (gssbzn) [reply test/e2e/atlas/helper_test.go:786]: yes

File: internal/store/data_federation.go
Comment (cveticm) [reply internal/store/data_federation.go:18]: Old SDK was used as it seemed to be standardly used in kubernetes side of the CLI and is used in Atlas K8s Operator.

Looking at the new SDK, it's api_data_federation uses DataLakeTenant but this seems to contain all same fields as the Data Federation stuct used by AKO, so fix should be relatively painless.

[This file](https://github.com/mongodb/mongodb-atlas-cli/blob/master/internal/store/atlas/data_federation.go) can be used as an alternative to store/data_federation. Fix will be needed to address that interface `AtlasOperatorGenericStore` cannot access methods in another package. store/atlas/data_federation will either have to move into store, or store/operator into store/atlas. There may be an alternative fix I'm not seeing.

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (cveticm) [inline internal/kubernetes/operator/datafederation/datafederation.go:37]: This method returns type mongodbatlas.DataFederationInstance. If store/data_federation.go will be changed to use [new SDK](https://github.com/mongodb/atlas-sdk-go), this method will return atlasv2.DataLakeTenant and all methods in this file will need to be adjusted accordingly.

File: test/e2e/atlas/helper_test.go
Comment (cveticm) [reply test/e2e/atlas/helper_test.go:786]: We may be able to avoid creating our own client to create a dataFed (as we are now) by using `atlas dataFederation create` command. This method was avoided since the command was thought to be non-functional. If flag `--file` is used, a json file with the desired specs for testing should work. 
Alternatively, the env var could be bypassed to allow for the rest of the code to remain as is. I'm unsure how this could be done so I will have to defer to @igor-karpukhin @helderjs

File: internal/kubernetes/operator/config_exporter.go
Comment (josvazg) [reply internal/kubernetes/operator/config_exporter.go]: Taking over the PR

Done

File: internal/store/data_federation.go
Comment (josvazg) [reply internal/store/data_federation.go:18]: investigating the SDK change

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (josvazg) [reply internal/kubernetes/operator/datafederation/datafederation.go:37]: Done

File: internal/store/data_federation.go
Comment (josvazg) [reply internal/store/data_federation.go]: Not sure if the evergreen issues are related to this changes or not.

**Comment by josvazg** [comment]
Down to 1 commit, to ease the rebase.

**Comment by josvazg** [comment]
Nope, reverting...

**Comment by josvazg** [comment]
I am not rebasing, rather will apply the exact same changes in this PR to the latest master HEAD.

File: test/e2e/atlas/helper_test.go
Comment (josvazg) [reply test/e2e/atlas/helper_test.go:824]: Got some hints from Igor on this, will try to fix in the next update.

Comment (josvazg) [reply test/e2e/atlas/helper_test.go:824]: I saw a green in evergreen. Closing for now.

File: internal/store/data_federation.go
Comment (josvazg) [reply internal/store/data_federation.go]: I saw a green in evergreen. Closing for now.

File: test/e2e/atlas/helper_test.go
Comment (josvazg) [reply test/e2e/atlas/helper_test.go]: Sent a possible fix. Not sure is 100% right as the original had more information that I could not make pass through the CLI interface for create data federated name. We'll see in the tests.

Comment (josvazg) [reply test/e2e/atlas/helper_test.go]: will comment on specific issues now

File: test/e2e/atlas/helper_test.go
Comment (gssbzn) [reply test/e2e/atlas/helper_test.go:791]: have you chekced if this task group exports the env variable on the evg config https://github.com/mongodb/mongodb-atlas-cli/blob/b55f413fe15afaac557f21ab06ce2aa7fdc454d6/build/ci/evergreen.yml

Comment (josvazg) [reply test/e2e/atlas/helper_test.go:791]: Thanks for the hint!

**Comment by josvazg** [comment]
@gssbzn please take a look when you can:

- It is migrated to the new API DSK.
- I fixed the test to use the atlas CLI rather than calling the API.
- It is rebased (or rather manually rewritten) on a recent change.

**Review by igor-karpukhin — COMMENTED** [review]
Looks good to me. @gssbzn please take a look when you have some time this week. Thanks in advance.

File: internal/kubernetes/operator/config_exporter_test.go
Comment (gssbzn) [inline internal/kubernetes/operator/config_exporter_test.go:1]:
```suggestion
// Copyright 2023 MongoDB Inc
```

File: internal/kubernetes/operator/config_exporter_test.go
Comment (gssbzn) [inline internal/kubernetes/operator/config_exporter_test.go:81]:
```suggestion
			nil, // credsProvider (not used)
```

File: internal/kubernetes/operator/config_exporter_test.go
Comment (gssbzn) [inline internal/kubernetes/operator/config_exporter_test.go:82]:
```suggestion
			projectID, "" // orgID (not used))
```

File: internal/kubernetes/operator/config_exporter_test.go
Comment (gssbzn) [inline internal/kubernetes/operator/config_exporter_test.go:93]:
```suggestion
		if diff := deep.Equal(result, expected); diff != nil {
			t.Error(diff)
		}
```

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation.go:71]:
```suggestion
	return state != DeletingState && state != DeletedState 
```

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation.go:96]: does this need to be exported?
```suggestion
func getDataProcessRegion(dataProcessRegion *admin.DataLakeDataProcessRegion) *atlasV1.DataProcessRegion {
```

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation.go:106]:
```suggestion
func getStorage(storage *admin.DataLakeStorage) *atlasV1.Storage {
```

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation.go:115]:
```suggestion
func getDatabases(database []admin.DataLakeDatabaseInstance) []atlasV1.Database {
```

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation.go:121]: you can probably preallocate this array

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation.go:137]: you can preallocate this array

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation.go:152]: save yourself a lib you don't need
```suggestion
			AllowInsecure:       obj.GetAllowInsecure(),
```

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation.go:194]:
```suggestion
			IncludeTags:              obj.GetIncludeTags(),
```

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation.go:196]:
```suggestion
			Public:                   obj.GetPublic(),
```

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation.go:188]: prealloc

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation.go:167]:
```suggestion
func getViews(views []admin.DataLakeApiBase) []atlasV1.View {
```

File: internal/kubernetes/operator/datafederation/datafederation.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation.go:182]:
```suggestion
func getStores(stores []admin.DataLakeStoreSettings) []atlasV1.Store {
```

File: internal/kubernetes/operator/datafederation/datafederation_test.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation_test.go:1]:
```suggestion
// Copyright 2023 MongoDB Inc
```

File: internal/kubernetes/operator/datafederation/datafederation_test.go
Comment (gssbzn) [inline internal/kubernetes/operator/datafederation/datafederation_test.go:209]:
```suggestion
		if diff := deep.Equal(result, expected); diff != nil {
			t.Error(diff)
		}

```

**Review by gssbzn — APPROVED** [review]
Thanks you all for the work here, good to merge if [atlas_kubernetes_e2e
Will](https://spruce.mongodb.com/task/mongodb_atlas_cli_master_e2e_atlas_kubernetes_atlas_kubernetes_e2e_patch_0bf1764552066d7e68dd209b93447ecc31a7dcf9_64f70ce97742aeedbe7ecde4_23_09_05_11_11_39/logs?execution=0) is green

**Comment by gssbzn** [comment]
evergreen keep-definitions

**Comment by igor-karpukhin** [comment]
K8S tests are all green. Merging.

**Pull request merged**
